File: java-workshop/part-2-light-mocks/src/test/java/service/DiscountCodeServiceTest.java
package service;

import domain.DiscountCode;
import org.junit.jupiter.api.Test;
import repository.DiscountCodeRepository;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

class DiscountCodeServiceTest {

    @Test
    void checkDiscountCode_returnsTrue_whenCodeHasAlreadyBeenUsed() {
        DiscountCodeRepository repository = mock(DiscountCodeRepository.class);
        DiscountCodeService discountCodeService = new DiscountCodeService(repository);

        when(repository.hasBeenUsed("CUST-001", DiscountCode.SUMMER20)).thenReturn(true);

        assertTrue(discountCodeService.checkDiscountCode("CUST-001", DiscountCode.SUMMER20));
    }

    @Test
    void checkDiscountCode_returnsFalse_whenCodeHasNotBeenUsed() {
        DiscountCodeRepository repository = mock(DiscountCodeRepository.class);
        DiscountCodeService discountCodeService = new DiscountCodeService(repository);

        when(repository.hasBeenUsed("CUST-001", DiscountCode.SUMMER20)).thenReturn(false);

        assertFalse(discountCodeService.checkDiscountCode("CUST-001", DiscountCode.SUMMER20));
    }

    @Test
    void markAsUsed_delegatesToRepository() {
        DiscountCodeRepository repository = mock(DiscountCodeRepository.class);
        DiscountCodeService discountCodeService = new DiscountCodeService(repository);

        discountCodeService.markAsUsed("CUST-001", DiscountCode.SUMMER20);

        verify(repository).markAsUsed("CUST-001", DiscountCode.SUMMER20);
    }
}

File: typescript-workshop/part-1-module-mocks/tests/service/DiscountCodeServiceTest.ts
// DiscountCodeService creates its DiscountCodeRepository inline (no DI).
// To test it in isolation we must mock BOTH the DiscountCodeService module
// (for use in OrderServiceTest) and the DiscountCodeRepository module.
//
// This test shows how to mock the nested dependency.

vi.mock('../../src/repository/DiscountCodeRepository');

import { describe, it, expect, vi } from 'vitest';
import { DiscountCode } from '../../src/domain/DiscountCode';
import { DiscountCodeRepository } from '../../src/repository/DiscountCodeRepository';
import { DiscountCodeService } from '../../src/service/DiscountCodeService';

describe('DiscountCodeService', () => {
  it('p1_checkDiscountCode_returnsTrue_whenCodeHasAlreadyBeenUsed', () => {
    // The DiscountCodeRepository constructor is mocked, so `new DiscountCodeRepository()`
    // inside DiscountCodeService returns a mock instance.
    vi.mocked(DiscountCodeRepository.prototype.checkDiscountCode).mockReturnValue(true);

    const service = new DiscountCodeService();
    expect(service.checkDiscountCode('CUST-001', DiscountCode.SUMMER20)).toBe(true);
  });

  it('p1_checkDiscountCode_returnsFalse_whenCodeHasNotBeenUsed', () => {
    vi.mocked(DiscountCodeRepository.prototype.checkDiscountCode).mockReturnValue(false);

    const service = new DiscountCodeService();
    expect(service.checkDiscountCode('CUST-001', DiscountCode.SUMMER20)).toBe(false);
  });

  it('p1_markAsUsed_delegatesToRepository', () => {
    const service = new DiscountCodeService();
    service.markAsUsed('CUST-001', DiscountCode.SUMMER20);

    expect(vi.mocked(DiscountCodeRepository.prototype.markAsUsed))
      .toHaveBeenCalledWith('CUST-001', DiscountCode.SUMMER20);
  });
});
